Exceptions not raised from class-based fixtures

Very useful library! I did find a bug with flaky, which is that if a class-based fixtures raises an exception, the exception is not raised, and for any tests that use that exception, it appears if the fixture never ran or exists. This makes it very hard to debug the fixture. (Originally I didn't know why this was happening, so I started uninstalling packages until I removed flaky, and things started working again)

(Actually I don't think it's just limited class-based fixtures, this happens in other cases too)

#### Test Example 1 (does a divide by zero to illustrate a bad fixture)
```python
import pytest

class TestStuff(object):
    @pytest.fixture(autouse=True)
    def setup(self):
        5 / 0
        self.foo = 'foo'

    def test_stuff(self):
        assert self.foo == 'foo'
```

Resulting output:
```bash
test_flaky.py F

=================================================== FAILURES ====================================================
_____________________________________________ TestStuff.test_stuff ______________________________________________

self = <test_flaky.TestStuff object at 0x10aac3d50>

    def test_stuff(self):
>       assert self.foo == 'foo'
E       AttributeError: 'TestStuff' object has no attribute 'foo'
```

#### Test Example 2 (variation of above, without autouse=True)
```python
class TestTwo(object):
    @pytest.fixture
    def setup(self):
        5 / 0
        self.foo = 'foo'

    def test_two(self, setup):
        assert self.foo == 'foo'
```

Resulting output:
```bash
self = <CallInfo when='call' exception: 'setup'>, func = <function <lambda> at 0x101b84668>
plugin = <flaky.flaky_pytest_plugin.FlakyPlugin object at 0x101b2d1d0>

    def call(self, func, plugin):
        """
            Call the test function, handling success or failure.
            :param func:
                The test function to run.
            :type func:
                `callable`
            :param plugin:
                Plugin class for flaky that can handle test success or failure.
            :type plugin:
                :class: `FlakyPlugin`
            """
        is_call = self.when == 'call'
        try:
>           self.result = func()

venv/lib/python2.7/site-packages/flaky/flaky_pytest_plugin.py:356:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
venv/lib/python2.7/site-packages/flaky/flaky_pytest_plugin.py:181: in <lambda>
    lambda: ihook(item=item, **kwds),
venv/lib/python2.7/site-packages/_pytest/core.py:521: in __call__
    return self._docall(self.methods, kwargs)
venv/lib/python2.7/site-packages/_pytest/core.py:528: in _docall
    firstresult=self.firstresult).execute()
venv/lib/python2.7/site-packages/_pytest/core.py:393: in execute
    return wrapped_call(method(*args), self.execute)
venv/lib/python2.7/site-packages/_pytest/core.py:113: in wrapped_call
    return call_outcome.get_result()
venv/lib/python2.7/site-packages/_pytest/core.py:138: in get_result
    py.builtin._reraise(*ex)
venv/lib/python2.7/site-packages/_pytest/core.py:123: in __init__
    self.result = func()
venv/lib/python2.7/site-packages/_pytest/core.py:394: in execute
    res = method(*args)
venv/lib/python2.7/site-packages/_pytest/runner.py:90: in pytest_runtest_call
    item.runtest()
venv/lib/python2.7/site-packages/_pytest/python.py:1181: in runtest
    self.ihook.pytest_pyfunc_call(pyfuncitem=self)
venv/lib/python2.7/site-packages/_pytest/core.py:521: in __call__
    return self._docall(self.methods, kwargs)
venv/lib/python2.7/site-packages/_pytest/core.py:528: in _docall
    firstresult=self.firstresult).execute()
venv/lib/python2.7/site-packages/_pytest/core.py:394: in execute
    res = method(*args)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

pyfuncitem = <Function 'test_two'>

    @pytest.mark.trylast
    def pytest_pyfunc_call(pyfuncitem):
        testfunction = pyfuncitem.obj
        if pyfuncitem._isyieldedfunction():
            testfunction(*pyfuncitem._args)
        else:
            funcargs = pyfuncitem.funcargs
            testargs = {}
            for arg in pyfuncitem._fixtureinfo.argnames:
>               testargs[arg] = funcargs[arg]
E               KeyError: 'setup'
```

#### Expected Output:
```bash
self = <test_flaky.TestOne object at 0x104c9fa90>

    @pytest.fixture(autouse=True)
    def setup(self):
>       5 / 0
E       ZeroDivisionError: integer division or modulo by zero

test_flaky.py:7: ZeroDivisionError
```

Test file with these examples at https://gist.github.com/alvinchow86/82ff8671dd8db847e251.

This is with Python 2.7.9, running on OS X Yosemite. Can reproduce this with just these PIP dependencies.
```
flaky==2.2.0
py==1.4.30
pytest==2.7.2
```

--- a/HISTORY.rst
+++ b/HISTORY.rst
@@ -9,6 +9,11 @@ Upcoming
 - Added support and testing for Python 3.5
 - Fixed tests on Python 2.6 with latest version of py.test
 
+**Bugfixes**
+
+- Flaky will no longer swallow exceptions raised during pytest fixture setup.
+  This change is correct, but is a change in behavior.
+
 2.2.0 (2015-08-28)
 ++++++++++++++++++
 

--- a/flaky/flaky_pytest_plugin.py
+++ b/flaky/flaky_pytest_plugin.py
@@ -1,7 +1,6 @@
 # coding: utf-8
 
 from __future__ import unicode_literals
-from time import time
 import py
 
 # pylint:disable=import-error
@@ -312,15 +311,11 @@ class FlakyCallInfo(CallInfo):
         # pylint:disable=super-init-not-called
         #: context of invocation: one of "setup", "call",
         #: "teardown", "memocollect"
-        self.when = when
-        self.start = time()
         self._item = item
         self._want_rerun = []
         self.excinfo = None
-        try:
-            self.call(func, plugin)
-        finally:
-            self.stop = time()
+        from functools import partial
+        CallInfo.__init__(self, partial(self.call, func, plugin), when)
 
     def _handle_error(self, plugin):
         """
@@ -373,6 +368,8 @@ class FlakyCallInfo(CallInfo):
         except:
             if is_call:
                 self._handle_error(plugin)
+            else:
+                raise
         else:
             if is_call:
                 if self.excinfo is not None:

--- a/test/test_flaky_pytest_plugin.py
+++ b/test/test_flaky_pytest_plugin.py
@@ -210,6 +210,33 @@ def test_flaky_plugin_can_suppress_success_report(
     assert string_io.getvalue() == mock_io.getvalue()
 
 
+def test_flaky_plugin_raises_errors_in_fixture_setup(
+        flaky_test,
+        flaky_plugin,
+        string_io,
+        mock_io,
+):
+    """
+    Test for Issue #57 - fixtures which raise an error should show up as
+    test errors.
+
+    This test ensures that exceptions occurring when running a test
+    fixture are copied into the call info's excinfo field.
+    """
+    def error_raising_setup_function(item):
+        assert item is flaky_test
+        item.ran_setup = True
+        return 5 / 0
+
+    flaky()(flaky_test)
+    flaky_test.ihook = Mock()
+    flaky_test.ihook.pytest_runtest_setup = error_raising_setup_function
+    call_info = flaky_plugin.call_runtest_hook(flaky_test, 'setup')
+    assert flaky_test.ran_setup
+    assert string_io.getvalue() == mock_io.getvalue()
+    assert call_info.excinfo.type is ZeroDivisionError
+
+
 class TestFlakyPytestPlugin(object):
     _test_method_name = 'test_method'
 

